encoders/src/main/scala/au/csiro/pathling/encoders2/SerializerBuilder2.scala

```diff
@@ -0,0 +1,228 @@
+import ca.uhn.fhir.context._
+import org.apache.spark.sql.catalyst.expressions.objects.{Invoke, MapObjects, StaticInvoke}
+import org.apache.spark.sql.catalyst.expressions.{BoundReference, CreateNamedStruct, Expression, If, IsNull, Literal}
+import org.apache.spark.sql.types.{DataType, DataTypes, ObjectType}
+import org.apache.spark.unsafe.types.UTF8String
+import org.hl7.fhir.instance.model.api.{IBaseDatatype, IBaseResource}
+import org.hl7.fhir.utilities.xhtml.XhtmlNode
+
+/**
+ * The schema processor for building serializer expressions.
+ *
+ * @param expression       the starting (root) expression.
+ * @param dataTypeMappings data type mappins to use.
+ */
+private[encoders2] class SerializerBuilderProcessor(expression: Expression, val dataTypeMappings: DataTypeMappings) extends
+  SchemaProcessorWithTypeMappings[Expression, ExpressionWithName] {
+
+  override def buildValue(childDefinition: BaseRuntimeChildDefinition, elementDefinition: BaseRuntimeElementDefinition[_], elementName: String,
+                          compositeBuilder: (SchemaProcessor[Expression, ExpressionWithName], BaseRuntimeElementCompositeDefinition[_]) => Expression): Seq[ExpressionWithName] = {
+    // add custom encoder
+    val customEncoder = dataTypeMappings.customEncoder(elementDefinition, elementName)
+    val evaluator: (Expression => Expression) => Expression = if (isCollection(childDefinition)) {
+      MapObjects(_,
+        expression,
+        objectTypeFor(childDefinition))
+    } else {
+      encoder => encoder(expression)
+    }
+    customEncoder.map(_.customSerializer2(evaluator))
+      .getOrElse(super.buildValue(childDefinition, elementDefinition, elementName, compositeBuilder))
+  }
+
+  override def buildArrayValue(childDefinition: BaseRuntimeChildDefinition, elementDefinition: BaseRuntimeElementDefinition[_], elementName: String,
+                               compositeBuilder: (SchemaProcessor[Expression, (String, Expression)], BaseRuntimeElementCompositeDefinition[_]) => Expression): Expression = {
+    MapObjects(withExpression(_).buildSimpleValue(childDefinition, elementDefinition, elementName, compositeBuilder),
+      expression,
+      objectTypeFor(childDefinition))
+  }
+
+  override def buildElement(elementName: String, elementValue: Expression, definition: BaseRuntimeElementDefinition[_]): ExpressionWithName = {
+    // Named serializer
+    (elementName, elementValue)
+  }
+
+  override def buildPrimitiveDatatype(primitive: RuntimePrimitiveDatatypeDefinition): Expression = {
+    dataTypeMappings.primitiveEncoderExpression(expression, primitive)
+  }
+
+  override def buildPrimitiveDatatypeXhtmlHl7Org(xhtmlHl7Org: RuntimePrimitiveDatatypeXhtmlHl7OrgDefinition): Expression = {
+    dataTypeToUtf8Expr(expression)
+  }
+
+  override def buildComposite(definition: BaseRuntimeElementCompositeDefinition[_], fields: Seq[(String, Expression)]): Expression = {
+    // TODO: Fix so that it does not traverse the type
```

Comment:
Any further explanation of this? Is this something that needs to be resolved before we merge this, or does another ticket need to be created?

Reply:
Good point. I left a few improvements to be addressed post Extension implementation as I am not sure the moment what the best way to fix them is and I expecting the design to evolve a bit when the extension need to be incorporated. So they can stay as todo and I have created a ticked [#414] to address them (Post or together with Extension support)